### ISSUE TITLE
Candidate Selection

### ISSUE THREAD

**Issue by bradfox2** [Praneet9/Representation-Learning-for-Information-Extraction#23]
Have you experimented with altering the candidate selection process? 

I am interested in what occurs when the candidate selection process is simplified or removed entirely so that every possible candidate is evaluated.

**Comment by Praneet9** [comment]
I didn't try that as removing it would mean a lot of meaningless negative candidates.
For eg: for a date, a floating-point number or random text doesn't make sense. So I didn't try that.

**Comment by Praneet9** [comment]
@Neelesh1121  Can you please explain what you mean by
> we have to extract all the key value present in the snapshot.

There's no rule like that. In the above case, what are you trying to extract for the amount key?

**Comment by panwar2001** [comment]
@Praneet9  Hi , i am really confused about positional embeddings, Do we have to collect every  invoice relative positions of  neighbors in the training set , then train the model to generate embedding for a 2 D coordinates like embedding([3,4])= [2.34,3.43,2.34........]
Or is it done by training one invoice at a time, but the embeddings generated for a particular relative position would be different each time while training for new invoice.

**Comment by Praneet9** [comment]
@panwar2001 Can you elaborate with an example of what issue you are facing?

**Comment by panwar2001** [comment]
@Praneet9  The embeddings are generated for 2D positions of neighbors , example : embedding((3,4)) =  [2.34,3.43,2.34........]
How this model learns new embeddings each time? 
Like to create words embeddings the words are first encoded into numerical values then we have number of words in the vocabulary finally neural network used to generate embeddings. 
How does this work with positions? 

class NeighbourEmbedding(nn.Module):

    def __init__(self, vocab_size, dimension):
        super().__init__()

        self.word_embed = nn.Embedding(vocab_size, dimension)
        self.linear1 = nn.Linear(2, 128)
        self.linear2 = nn.Linear(128, dimension)
        self.dropout1 = nn.Dropout(0.5)
        self.dropout2 = nn.Dropout(0.5)

    def forward(self, words, positions):

        # Word embedding
        embedding = self.word_embed(words)

        # Position embedding
        pos = F.relu(self.linear1(positions))
        pos = self.dropout1(pos)
        pos = F.relu(self.linear2(pos))
        pos = self.dropout1(pos)

        # Concatenating word and position embeddings
        neighbour_embedding = torch.cat((embedding, pos), dim=2)

        return neighbour_embedding

**Comment by panwar2001** [comment]
@Praneet9  Hi, so it is written that N neigbour encodings are first projected from N*2d to 4* 2d * N  size , where d is embedding size and N is number of neigbours.Example d=128
**Further,    4* 2d* N encoding are projected down to 2*N encoding ,**  
How is it done?? 
Is it projected down using Dense layer or using MAX pooling layer. 
They have mentioned it's projected down , then later have mentioned that max pooling the neighbour encoding was found to be more useful ???
**why you have not projected down directly using max pooling and instead used dense layer first?**

![inv](https://github.com/Praneet9/Representation-Learning-for-Information-Extraction/assets/130741616/cf8661e9-b4d4-4278-b9d2-e47177d3d31c)

**Comment by Praneet9** [comment]
@panwar2001  I think you are confusing that projection with the neighbour encoding. The neighbour encodings are projected to 4 * 2d and then maxpooled as you can see [here](https://github.com/Praneet9/Representation-Learning-for-Information-Extraction/blob/b268463e312689e3dc5222cebf0f5a2e4be68fb6/network/model.py#L44). These tensors are then concatenated with candidate embeddings, which are then projected back down. Hope that answers your confusion.
